Hotfix: prevent http server loading on import

taylorzane opened the pull request:
This PR fixes a bug that I encountered, that will prevent a script from exiting on completion (e.g. gulp/grunt) if the module is imported (or is imported via a sub-module).

The script now checks to see if the program is being run by itself, and if so, immediately starts the server. If the module is imported, then it waits for the first `Subscription` instance to be created. Only one server will be started for all instances of `Subscription`.

taylorzane commented:
Let me know what you think about the implementation of this. I definitely think this is a required fix, but you may have some suggestions on how you'd like this to be implemented.

bazwilliams commented:
Good suggestion. Will have a look over this over the weekend. This came about due to a recent move away from an http server per subscription to a single one.

taylorzane commented:
I've got a couple of other ideas. But that was the one that fit most easily with your existing code. If you're concerned about one server per instance you could always create a singleton instance to return. 

We can certainly chat more this weekend. 

Sent from my iPhone

> On Sep 23, 2016, at 12:40, Barry Williams notifications@github.com wrote:
> 
> Good suggestion. Will have a look over this over the weekend. This came about due to a recent move away from an http server per subscription to a single one.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

bazwilliams commented:
By all means create issues for the other suggestions, this was written primarily as a refactor of common code from other applications I'd written so there will be many features I've not thought of. 

I've spiked out a version of the code which imports the httpserver through a `require`. There is a potential race condition so we'd need a callback or event so we don't initialise the subscription until the server is actually listening.

bazwilliams commented:
I'm not sure about the use case where the server starts if the script is run standalone, I'm thinking just initialise on first subscription regardless and would remove the `module.parent` checks. This might be why the tests are failing.

taylorzane commented:
You're right on the standalone initialization. I'll remove that and update the variable names :+1:

taylorzane commented:
I changed the variable names. But I'm still unable to get this test working. I think that it's failing because the subscription events are happening too quickly together (I just measured, and they're happening at the exact same millisecond). So either `sub1` or `sub2` isn't defined yet. I'm going to try to maybe delay the return to see if that helps. Async testing is such a PITA...

taylorzane commented:
@bazwilliams This is a pretty big change at this point... I've fixed all of your requests. 

I had to make the http server act asynchronously in order to prevent any subscribers from jumping the gun and passing an undefined port.

I also added a check to ensure that not only had the http server been requested to start, but was actually successfully started, and listening.

I fixed the test, I had to make the test asynchronous, using the `async` library. But it should make the test much more resilient.

I think that's all I can think of. I you have any questions in the code, don't hesitate to point it out and ask me about it!

bazwilliams commented on index.js:
```diff
@@ -7,33 +7,48 @@ let util = require('util');
+var httpServerEmitter = new events();
+var httpServerStarting = false;
+var httpServerStarted = false;
+var httpServerPort;
```
Minor thing - can you change these to 'let' please?

taylorzane replied:
I can certainly do that. I think `let` was causing problems with Node.js v6, I'll test and let you know. The issues I was encountering were probably due to the pre-async design of the http server initialization.

bazwilliams commented on tests/whenHandlingMultipleSubscriptions.js:
```diff
@@ -49,20 +51,27 @@ describe('When handling multiple subscriptions', function () {
-        subscription2.on('subscribed', function(payload) {
-            sid2 = payload.sid;
-            if (sub1) {
+        }, function(err, results) {
```
const xmlResponseParser = require('parsexmlresponse');

bazwilliams replied:
Nice! Really good idea to use `async` instead of my shoddy implementation before!

taylorzane commented:
Alright, I updated those for you. It didn't cause any issues with Node.js v6.

bazwilliams commented:
LGTM :+1:

bazwilliams commented:
I've packaged this up to npm. Thanks very much. Anything else, feel free to create issues/PRs!
